drivers: can: Xilinx Processor System CAN controller driver

ibirnbaum opened the pull request:
Driver implementation for the Xilinx Processor System CAN controller, which is contained in both the Zynq-7000 and ZynqMP (UltraScale) SoCs with two instances each.

Known limitations:
- The high priority TX FIFO is unsupported, as the CAN driver API doesn't seem to consider such a special type of TX FIFO.
- The send function is currently not SMP-safe if it gets called on 2 CPUs simultaneously and the function is called with no callback function pointer on either core. This should be tolerable for now as no SMP-capable Zynq/ZynqMP-based target currently exists.
- SocketCAN integration not yet tested.

Basic functional tests performed with a Zynq-based Zedboard with a Microchip CAN transceiver attached externally and a Raspberry Pi as peer system. Test cases tests/drivers/can/api and tests/drivers/can/timing passed in loopback mode on the Zedboard.

ibirnbaum committed:
drivers: can: Xilinx Processor System CAN controller

Initial commit of a driver implementation for the Xilinx Processor
System CAN controller.

- Device memory is mapped using the DEVICE_MMIO API
- I/O pin mapping via the Xilinx SLCR syscon is supported if enabled

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
dts: bindings: can: Xilinx Processor System CAN controller binding

DT binding for the Xilinx Processor System CAN controller.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
dts: arm: xilinx: DT data for Xilinx CAN controller instances

DT data for the instances of the Xilinx Processor System CAN
controller contained in both the Zynq-7000 and ZynqMP (UltraScale)
SoCs.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
CODEOWNERS: Xilinx Processor System CAN driver

set code ownership for /drivers/can/*xlnx*

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum commented:
Fixed extended ID handling, the test cases tests/drivers/can/api and tests/drivers/can/timing now pass on a Zynq-7000 based Zedboard.

henrikbrixandersen commented on dts/bindings/can/xlnx,can.yaml:
```diff
@@ -0,0 +1,32 @@
+  This CAN controller is contained in both the Xilinx Zynq-7000 and
+  ZynqMP (UltraScale) SoCs.
+
+compatible: "xlnx,can"
```
Please use the same compatible as used in the upstream Linux devicetree bindings ("xlnx,zynq-can-1.0"). See https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/net/can/xilinx_can.txt

henrikbrixandersen commented on dts/bindings/can/xlnx,can.yaml:
```diff
@@ -0,0 +1,32 @@
+    interrupts:
+      required: true
+
+    clock-frequency:
+      type: int
+      required: true
+      description: |
+        Specifies the base clock frequency from which the CAN controller's
+        clock frequency for the selected baud rate will be derived using
+        a divider in the respective CAN controller's Baud Rate Prescaler
+        Register. Therefore, the value of this item must be set to the clock
+        frequency of the PLL supplying the respective CAN controller's clock
+        - by default, this is the IO PLL.
```
Can't this frequency be established using the clock controller framework and devicetree?

henrikbrixandersen reviewed (changes requested):
Thank you for this addition. A couple of initial comments. I am still reviewing the driver itself.

I assume all the CAN test (`twister -p xxx -t can --device-testing ...`) pass? Have you verified that queued tx frames are sent in priority order (and not in FIFO order)?

ibirnbaum commented:
Hi @henrikbrixandersen,

I've re-based and updated the driver to match the recent changes in the API (timing config, operational modes, get_max_bitrate function etc.) and fixed the preservation of the current operational mode across a timing modification (this came up when the tests start off in loopback mode, but ended up in normal mode after a timing change, preventing them from completing). Is there a reference application to be used as peer for the isotp tests? Those tests do run, but without a valid peer do not produce the desired test results. canopennode.no_storage won't run due to a shortage of filters (13 required, 4 available, that's why I asked regarding the CAN_MAX_FILTER setting on Discord the other day), but drivers.can.api, drivers.can.timing, drivers.can.utilities and the socketcan samples work & pass.

ibirnbaum commented:
Changes from 08/08/22:
- driver now uses the DEVICE_MMIO API to obtain the current device's (virtual, if applicable) base address rather than just using the physical base address from the device tree.
- modified the Zynq/ZynqMP DTS files to no longer include a "label" property.

galak commented on drivers/can/Kconfig.xlnx:
```diff
@@ -0,0 +1,17 @@
+	bool "Xilinx Processor System CAN controller driver"
+	default $(dt_compat_enabled,$(DT_COMPAT_XLNX_CAN))
+	depends on SOC_XILINX_ZYNQMP_RPU || SOC_FAMILY_XILINX_ZYNQ7000
+	depends on !QEMU_TARGET
```
remove `DT_COMPAT_XLNX_CAN`

and make this:

```
default y
depends on DT_HAS_XLNX_ZYNQ_CAN_1_0_ENABLED
```

remove `depends on SOC_XILINX...`

galak reviewed (changes requested):
Update Kconfig for new DT_HAS_<COMPAT>_ENABLED support.

ibirnbaum commented:
> Update Kconfig for new DT_HAS__ENABLED support.

done.

ibirnbaum commented:
Changes from 08/08/22, part 2:
- changed Kconfig to new format
- added integration with the matching pinctrl driver for the Zynq SLCR.

ibirnbaum commented:
Changes from 11/08/22:
- removed deprecated DT_COMPAT_XLNX_CAN variable from Kconfig

ibirnbaum commented:
25.08.22:

- now using zephyr/ namespace for includes in can_xlnx.c
- changed zcan_ prefix to can_ after merge of #48941

henrikbrixandersen commented on drivers/can/can_xlnx.h:
```diff
@@ -0,0 +1,385 @@
+/* The Xilinx CAN controller always has 4 acceptance filters */
+#define CAN_XLNX_NUM_ACCEPTANCE_FILTERS	4
+/* The RX/TX FIFOs of the Xilinx CAN controller are always 64 entries deep */
+#define CAN_XLNX_RX_TX_FIFO_DEPTH	64
```
I'm disappointed to see (from the Z7K TRM) that the Xilinx PS7 CAN IP TX FIFO suffers from priority inversion. Even though having a nice HW FIFO available, it's no good when its sorting does not adhere to the priority of the CAN frames contained in it (it would need a HW priority queue instead).

I'm afraid using the HW FIFO here is not acceptable as it can lead to a high-priority frame having to wait for up to 63 lower-priority frames to be sent before itself.

I am slowly making progress on a software priority queue for use in CAN drivers where the underlying hardware has this "feature" (or misses this feature, has this bug, ...), but it is not ready yet. For now I suggest limiting this driver to use only _one_ position in the HW TX FIFO at a time. Sorry.

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+ * Copyright (c) 2022, Weidmueller Interface GmbH & Co. KG
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Known limitations / TODOs:
+ * - Sleep mode / PM functionality not considered at the time being.
+ * - High Priority TX buffer not considered at the time being.
```
Please leave out TODOs from the code. From the code it can easily be seen that PM functionality is not implemented - and the high-priority TX buffer does not fit into the Zephyr CAN API and likely never will.

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+#include "can_xlnx.h"
+#include "can_utils.h"
+
+#define LOG_MODULE_NAME can_xlnx
+#define LOG_LEVEL CONFIG_CAN_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
```
Please just provide these directly to the `LOG_MODULE_REGISTER()` call.

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+		}
+
+		/* Execute state change callback if registered */
+		if (dev_data->state_chg_cb != NULL) {
+			LOG_DBG("%s calling state change callback function",
+				dev->name);
+
+			error_counters.rx_err_cnt = dev_data->rx_errors;
+			error_counters.tx_err_cnt = dev_data->tx_errors;
+			dev_data->state_chg_cb(dev, state, error_counters,
+					       dev_data->state_chg_cb_data);
```
You need to make a copy of `dev_data->state_chg_cb`, check it against `NULL` and if non-NULL, call the copy. This code may lead to the `dev_data->state_chg_cb` value being changed between checking it and using it.

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+	 * Bus off state can not be detected via the error counters, entering this
+	 * error state is indicated by a separate interrupt flag.
+	 *
+	 * NOTICE: this could alternatively be determined by evaluating the ESTAT
+	 * bits in the Status Register (SR). Comp. Zynq-7000 TRM, register SR de-
+	 * scription, Appendix B.5, p. 804), but we need the counters updated any-
+	 * ways as we need up-to-date values whenever the counters are queried
+	 * via can_xlnx_get_state.
```
Or you could just read out the counter values in `can_xlnx_get_state()`. I would prefer determining the error state using the state itself and not the error counters.

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+					continue;
+				}
+
+				if (!can_utils_filter_match(&rx_frame,
```
Looks like you are mixing HW RX filtering with SW RX filtering here?

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+				dev_data->rx_filters[i].cb(dev, &rx_frame,
+							   dev_data->rx_filters[i].cb_arg);
+			}
+		} else {
+			LOG_DBG("%s discarded received frame with ID 0x%X as no "
+				"RX filters are registered", dev->name, rx_frame.id);
```
Please leave this out as it will make it almost impossible to debug anything on a loaded CAN bus.

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+		if (isr & CAN_XLNX_IRQ_RX_UNDERFLOW) {
+			sys_write32(CAN_XLNX_IRQ_RX_UNDERFLOW,
+				    base_addr + CAN_XLNX_ICR_OFFSET);
+			LOG_ERR("%s read from RX FIFO caused an undeflow error",
```
Should this be `LOG_WRN()`? The issue seem to be recoverable?

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+	/* Clear the TX OK interrupt bit */
+	sys_write32(CAN_XLNX_IRQ_MESSAGE_TX, base_addr + CAN_XLNX_ICR_OFFSET);
+
+	/* Process all pending TX done handlers */
```
Why process all pending TX done "handlers" here? Only the callbacks for the frames actually transmitted are to be called.

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+	 * them: comp. Zynq-7000 TRM, chap. 18.2.1.
+	 */
+
+#ifdef CONFIG_CAN_FD_MODE
+	if (mode & CAN_MODE_FD) {
+		LOG_ERR("%s CAN FD is unsupported", dev->name);
+		return -ENOTSUP;
+	}
+#endif
```
No need for the `#ifdef` here. This should always produce an error regardless of the Kconfig.

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+	}
+
+	/* Configure operational mode in the Mode Select Register */
+	switch (mode) {
+	case CAN_MODE_NORMAL:
+		/* No explicit indication of normal mode */
+		break;
+	case CAN_MODE_LISTENONLY:
+		msr |= CAN_XLNX_MSR_SNOOP;
+		break;
+	case CAN_MODE_LOOPBACK:
+		msr |= CAN_XLNX_MSR_LOOPBACK;
+		break;
+	default:
+		/* CAN_SILENT_LOOPBACK_MODE: unsupported by the controller */
+		return -EINVAL;
+	};
```
Doing this a switch statement makes assumptions about the values of the `CAN_MODE_` definitions. `mode` is a bitfield and must be treated as such. See the other drivers for inspiration.

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+		default:
+			break;
+		};
+	} while (!target_mode_reached);
```
Could be an endless loop? How come the loop is needed?

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+
+	/* Cannot send if the controller is in bus off state */
+	if (dev_data->bus_off == 1) {
+		return -ENETDOWN;
```
Needs rebasing and change after fda3f54002408f7cb5a28285b89ce3fd9300f238.

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+
+	/* Check if the current operational mode is suitable for transmission */
+	sr = sys_read32(base_addr + CAN_XLNX_SR_OFFSET);
+	if (sr & (CAN_XLNX_SR_CONFIG_MODE | CAN_XLNX_SR_SNOOP_MODE)) {
+		LOG_ERR("Cannot send CAN frame via interface %s: unable to transmit "
+			"in %s mode", dev->name, (sr & CAN_XLNX_SR_CONFIG_MODE) ?
+			"config" : "snoop");
+		return -EIO;
+	}
```
Listen-only mode should not make the `can_send()` API call fail. What gave you the impression it should?

ibirnbaum replied:
If send is called while in listen-only mode and a callback is provided, should it be called before returning or should it be ignored?

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+
+	/* Cannot send more frames if the TX FIFO is full */
+	isr = sys_read32(base_addr + CAN_XLNX_ISR_OFFSET);
+	if (isr & CAN_XLNX_IRQ_TX_FULL) {
+		LOG_ERR("Cannot send CAN frame via interface %s: TX FIFO full", dev->name);
+		return -ENOSPC;
+	}
```
You need to honor the timeout for acquiring a free TX mailbox/FIFO slot here instead of just failing. But since the driver needs to be migrated to only using one TX FIFO slot anyways, this needs to rewritten.

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+	/* Wait for filter to be idle */
+	do {
+		sr = sys_read32(base_addr + CAN_XLNX_SR_OFFSET);
+	} while (sr & CAN_XLNX_SR_ACC_FLTR_BUSY);
```
Another loop which can potentially block forever.

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+		 */
+		can_xlnx_state_update(dev, CAN_BUS_OFF);
+#ifdef CONFIG_CAN_AUTO_BUS_OFF_RECOVERY
+		/*
+		 * The controller supports auto-recovery. In order to determine
+		 * when auto-recovery has completed, poll SR[ESTAT] (comp. Zynq-
+		 * 7000 TRM, Appendix B.5, p. 804) for any value other than 0x2
+		 * (indicates bus off).
+		 */
+		do {
+			sr = sys_read32(base_addr + CAN_XLNX_SR_OFFSET);
+		} while (((sr >> CAN_XLNX_SR_ERROR_STATUS_OFFSET) &
+			CAN_XLNX_SR_ERROR_STATUS_MASK) == CAN_XLNX_SR_ESTAT_BUS_OFF);
+
+		can_xlnx_state_update(dev, CAN_ERROR_ACTIVE);
```
Polling in an interrupt handler?

henrikbrixandersen commented on drivers/can/can_xlnx.c:
```diff
@@ -0,0 +1,1329 @@
+ *
+ * @param dev Pointer to the CAN controller's device.
+ */
+static void can_xlnx_configure_irq(const struct device *dev)
+{
+	const struct can_xlnx_dev_cfg *dev_conf = dev->config;
+
+	/* Attach the ISR to the respective controller's IRQ */
+	dev_conf->config_func(dev);
+}
```
Why an entire function just for calling a function pointer?

ibirnbaum replied:
resolved, inlined into the init function

henrikbrixandersen commented:
@ibirnbaum You'll need to rebase this and add support for stopping/starting the CAN controller to comply with the recent CAN controller API changes.

ibirnbaum commented:
@henrikbrixandersen Thanks for the review, started working on an overhaul of the driver, more comments to some of your individual points coming up.

ibirnbaum commented on drivers/can/can_xlnx.h:
```diff
@@ -0,0 +1,385 @@
+/* Device config & run-time data struct creation macros */
+#define CAN_XLNX_DEV_DATA(idx)\
+static struct can_xlnx_dev_data can_xlnx##idx##_data = {\
+	.state = CAN_ERROR_ACTIVE,\
```
done, now using the new CAN_STATE_... definitions.

ibirnbaum commented on drivers/can/can_xlnx.h:
```diff
@@ -0,0 +1,385 @@
+#define CAN_XLNX_DEV_DEFINE(idx)\
+DEVICE_DT_INST_DEFINE(idx, can_xlnx_init, NULL,\
+	&can_xlnx##idx##_data, &can_xlnx##idx##_cfg,\
+	PRE_KERNEL_2, CONFIG_CAN_INIT_PRIORITY, &can_xlnx_apis);
```
Changed. I'm not sure any more where this originally came from. Most likely from the assumption that devices in general should be ready before the Kernel.

ibirnbaum commented on drivers/can/can_xlnx.h:
```diff
@@ -0,0 +1,385 @@
+ */
+struct can_xlnx_filter_data {
+	can_rx_callback_t cb;
+	void *cb_arg;
```
done

ibirnbaum commented on drivers/can/can_xlnx.h:
```diff
@@ -0,0 +1,385 @@
+	DEVICE_MMIO_RAM;
+	struct k_sem filter_prot_sem;
+	struct k_sem tx_done_prot_sem;
+	struct k_sem tx_done_sem; /* TODO: array with size [num_cpus] */
```
I'm still targeting SMP support for the Zynq and the R-cores of the UltraScale. Should we eventually get there, this driver, the existing GEM driver as well as any future Xilinx PS peripheral drivers must consider the possibility that (usually TX) functions that perform a blocking wait which is ended in an interrupt-driven fashion via a semaphore (or mutex) could be called on both CPU cores simultaneously. In that case, individual semaphores for both of those calls are required. Now that k_event has been recently introduced, that might be an alternative road to go down as well - replacing one or more semaphores with a single k_event in which each bit represents an individual CPU core.

henrikbrixandersen commented:
@ibirnbaum This has been open for 1+ year already. Either leave it closed, convert to draft, or follow up on the review comments please.
